keps/sig-auth/20190607-certificates-api.md

```diff
@@ -0,0 +1,285 @@
+Kubelet API because verifying IP SANs for Kubelets in a generic way poses
+challenges.
+
+The GCP controller manager replaces the in-built approver with an
+[approver](https://github.com/kubernetes/cloud-provider-gcp/blob/08fa1e3260ffb267682762e24ba93692000e3be8/cmd/gcp-controller-manager/csr_approver.go)
+that handles TLS certificates for server authentication and support for
+verification of CSRs with proof of control of GCE vTPM.
+
+An external project, [kapprover](https://github.com/coreos/kapprover), does
+policy based approval of kubelet CSRs.
+
+<!--
+TODO: various things use the certificates API for stuff other than Kubelet TLS
+certficates. For example:
+
+* Istio
+* CockroachDB
+
+Investigate these usages and report back.
+-->
+
+### Manual CSR Approval
+
+A Kubernetes administrator (with appropriate permissions) can manually approve
+(or deny) Certificate Signing Requests by using the `kubectl certificate
+approve` and `kubectl certificate deny` commands.
+
+### Automatic Signers
+
+The kube-controller-manager ships with an in-built
+[signer](https://github.com/kubernetes/kubernetes/blob/32ec6c212ec9415f604ffc1f4c1f29b782968ff1/pkg/controller/certificates/signer/cfssl_signer.go)
+that signs all approved certificates with a local key. This key is generally
+configured to be the key of the cluster's root certificate authority, but is not
+required to be.
+
+The GCP controller manager implements a
+[signer](https://github.com/kubernetes/cloud-provider-gcp/blob/08fa1e3260ffb267682762e24ba93692000e3be8/cmd/gcp-controller-manager/csr_signer.go)
+that uses a webhook to sign all approved CSRs. This allows the root certificate
+authority secret material to be stored and maintained outside of the Kubernetes
+control plane.
+
+<!--
+### Certificates API Clients
+
+TODO: talk about what's provided by k8s.io/client-go and what other people have
+built, e.g.
+[certificate-init-container](https://github.com/kelseyhightower/certificate-init-container).
+-->
+
+## Graduation Criteria
+
+### Beta -> GA Graduation
+
+- TBD
+
+## Implementation History
+
+- The Certificates API was merged as Alpha in 1.4
+- The Certificates API was promoted to Beta in 1.6
+
+## Known Issues
+
+- API shape/issues
+  - Requires requesters to know all the info about the end certificate.
+  - Use for higher-level requests (“give me a node client cert”, “give me a
+    serving cert for a pod backing a service”).
+  - Requested certificate attributes are split unthoughfully between encoded CSR
+    and fields in request spec which create difference in semantics.
+- Approval flow/issues
+  - Cannot limit or add components to request (limit or add SANs, usages, etc)
+- Signing flow/issues
+  - Method for multiple signers to interact (or approver to indicate what signer
+    should be used)
+- Guarantees on issued certificates
+  - No (current) guarantee all requested extensions/SANs are issued
+  - No (current) guarantee issued client certificates will be accepted as API
```

Comment:
That's true, but:
- why is it a problem?
- would it apply for custom Certificates API usages (other than kubelet certs)?

Reply:
Removed this section.